### PR TITLE
xds: Verify that the ServiceAccount from the NodeID is the same as the one in the mTLS cert's CN

### PR DESCRIPTION
**Pull request by draychev**
We encode Kubernetes the Service Account of the Pod on which a given Envoy is in 2 places:
1. the Node ID of the Envoy
2. in the CN of the xDS certificate in the bootstrap config of the Envoy

This PR adds validation at gRPC connect (Envoy to ADS).  gRPC will be disconnected if the Service Account in NodeID is not the same as the one in the mTLS certificate.

### CODE REVIEW

File: pkg/envoy/ads/grpc.go [hunk 1]
```diff
@@ -28,26 +30,45 @@ func receive(requests chan xds_discovery.DiscoveryRequest, server *xds_discovery
...
 
+			// Verify that the ServiceAccount from the NodeID is the same as the one in the mTLS cert's CN
+			cn := proxy.GetCertificateCommonName()
+			certSA, err := catalog.GetServiceAccountFromProxyCertificate(cn)
+			if err != nil {
+				log.Err(err).Msgf("Error getting service account from XDS certificate with CommonName=%s", cn)
+				return err
+			}
+
+			nodeIDSA := fmt.Sprintf("%s/%s", meta.Namespace, meta.ServiceAccount)
```
Comment (shashankram) [inline pkg/envoy/ads/grpc.go:60]: nit: s/nodeIDSA/svcAccFromNodeID, s/certSA/svcAccFromCert

File: pkg/envoy/ads/grpc.go [hunk 2]
```diff
@@ -28,26 +30,45 @@ func receive(requests chan xds_discovery.DiscoveryRequest, server *xds_discovery
...
+				return
+			}
 		}
 		log.Trace().Msgf("[grpc] Received DiscoveryRequest from Envoy with certificate SerialNumber %s", proxy.GetCertificateSerialNumber())
 		requests <- *request
 	}
 }
 
-func recordEnvoyPodMetadata(request *xds_discovery.DiscoveryRequest, proxy *envoy.Proxy, proxyRegistry *registry.ProxyRegistry) {
+func recordEnvoyPodMetadata(request *xds_discovery.DiscoveryRequest, proxy *envoy.Proxy, proxyRegistry *registry.ProxyRegistry) error {
```
Comment (shashankram) [inline pkg/envoy/ads/grpc.go:44]: I'd add a test for this function and the newly added code.

**Review by shashankram — APPROVED** [review]

**Review by eduser25 — APPROVED** [review]
This is a good change.
If someone is however able to steal a cert, I imagine it should not be hard for them to also launch an envoy and modify the node_id on envoy bootstrap I imagine

File: pkg/envoy/ads/grpc.go (hunk 2 above)
Comment (eduser25) [reply pkg/envoy/ads/grpc.go:43]: +1

Comment (draychev) [reply pkg/envoy/ads/grpc.go:43]: Made https://github.com/openservicemesh/osm/issues/3363 for this - to follow up shortly.

**Pull request merged**